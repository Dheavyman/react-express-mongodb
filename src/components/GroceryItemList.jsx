import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';

import GroceryItem from './GroceryItem';
import GroceryListAddItem from './GroceryListAddItem';
import * as groceryActions from '../actions';

const propTypes = {
  items: PropTypes.arrayOf(PropTypes.shape({
    name: PropTypes.string,
    purchased: PropTypes.bool,
  })).isRequired,
  actions: PropTypes.shape({
    addGroceryItem: PropTypes.func.isRequired
  }).isRequired
};


/**
 * Grocery item list component
 *
 * @class GroceryItemList
 * @extends {Component}
 */
class GroceryItemList extends Component {
  /**
   * Creates an instance of GroceryItemList.
   *
   * @param {object} props - Properties passed to component
   * @memberof GroceryItemList
   */
  constructor(props) {
    super(props);
    this.state = {
      input: ''
    };
    this.handleChangeInputName = this.handleChangeInputName.bind(this);
    this.addItem = this.addItem.bind(this);
  }

  /**
   * Change input name handler
   *
   * @param {object} event - Event object
   *
   * @returns {any} Set new state
   * @memberof GroceryItemList
   */
  handleChangeInputName(event) {
    this.setState({
      input: event.target.value
    });
  }

  /**
   * Add item handler
   *
   * @param {object} event - Event object
   *
   * @returns {any} Call add grocery item action and set new state
   * @memberof GroceryItemList
   */
  addItem(event) {
    event.preventDefault();
    const { input } = this.state;
    const { actions } = this.props;

    if (input.trim() !== '') {
      actions.addGroceryItem({
        name: input.trim()
      });
    };
    this.setState({
      input: ''
    });
  }

  /**
   * Render function
   *
   * @returns {object} React element
   * @memberof GroceryItemList
   */
  render() {
    const { input } = this.state;
    const { items, actions } = this.props;

    return (
      <div>
        <div className="card my-4 pl-3">
          <h2 className="header">Grocery Listify</h2>
        </div>
        <div className="row pl-3">
          {items.map((item, index) => (
            <GroceryItem
              key={`item ${index.toString()}`}
              item={item}
              actions={actions}
            />
          ))}
        </div>
        <div className="row">
          <GroceryListAddItem
            input={input}
            handleChangeInputName={this.handleChangeInputName}
            addItem={this.addItem}
          />
        </div>
      </div>
    );
  }
}

GroceryItemList.propTypes = propTypes;

/**
 * Map state to props
 *
 * @param {object} state - State
 *
 * @returns {object} Stated mapped to props
 */
const mapStateToProps = state => ({
  items: state.items
});

/**
 * Map dispatch to props
 *
 * @param {function} dispatch - Dispatch
 *
 * @returns {object} Dispatch mapped to props with actions
 */
const mapDispatchToProps = dispatch => ({
  actions: bindActionCreators(groceryActions, dispatch)
});

export default connect(mapStateToProps, mapDispatchToProps)(GroceryItemList);
