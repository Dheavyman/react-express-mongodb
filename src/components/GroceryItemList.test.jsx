import React from 'react';
import TestRenderer from 'react-test-renderer';

import { GroceryItemList } from './GroceryItemList';

describe('GroceryItemList component', () => {
  let props;
  beforeEach(() => {
    props = {
      items: [{
        name: 'Pizza',
        purchased: false
      }, {
        name: 'Malt',
        purchased: false
      }],
      actions: {
        addGroceryItem: jest.fn(),
        updateGroceryItem: jest.fn(),
        deleteGroceryItem: jest.fn()
      }
    };
  });

  it('should render component', () => {
    const renderer = TestRenderer.create(<GroceryItemList {...props} />);

    expect(renderer.toJSON()).toMatchSnapshot();
  });
});
