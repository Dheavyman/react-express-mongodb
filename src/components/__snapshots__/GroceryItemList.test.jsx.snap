// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GroceryItemList component should render component 1`] = `
<div>
  <div
    className="card header my-4 pl-3"
  >
    <h2>
      Grocery Listify
    </h2>
  </div>
  <div
    className="row pl-3"
  >
    <div
      className="col-sm-12"
    >
      <div
        className="row"
      >
        <div
          className="col-5 col-lg-6"
        >
          <h4
            className="grocery-item"
          >
            Pizza
          </h4>
        </div>
        <div
          className="col-3 col-lg-2"
        >
          <form
            className="three-columns"
            onSubmit={[Function]}
          >
            <button
              className="btn btn-primary"
              type="submit"
            >
              Buy
            </button>
          </form>
        </div>
        <div
          className="col-4 col-lg-4"
        >
          <form
            onSubmit={[Function]}
          >
            <button
              className="btn btn-outline-danger"
              type="submit"
            >
              ×
            </button>
          </form>
        </div>
      </div>
      <hr />
    </div>
    <div
      className="col-sm-12"
    >
      <div
        className="row"
      >
        <div
          className="col-5 col-lg-6"
        >
          <h4
            className="grocery-item"
          >
            Malt
          </h4>
        </div>
        <div
          className="col-3 col-lg-2"
        >
          <form
            className="three-columns"
            onSubmit={[Function]}
          >
            <button
              className="btn btn-primary"
              type="submit"
            >
              Buy
            </button>
          </form>
        </div>
        <div
          className="col-4 col-lg-4"
        >
          <form
            onSubmit={[Function]}
          >
            <button
              className="btn btn-outline-danger"
              type="submit"
            >
              ×
            </button>
          </form>
        </div>
      </div>
      <hr />
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="col-sm pl-4"
    >
      <form
        className="form-inline"
        onSubmit={[Function]}
      >
        <div
          className="form-group mt-3 mt-sm-0"
        >
          <label
            className="sr-only"
            htmlFor="inputGroceryItem"
          >
            Grocery item
          </label>
          <input
            className="form-control"
            id="inputGroceryItem"
            onChange={[Function]}
            placeholder="Grocery item"
            type="text"
            value=""
          />
        </div>
        <button
          className="btn btn-outline-primary"
          type="submit"
        >
          Add Item
        </button>
      </form>
    </div>
  </div>
</div>
`;
